precompiles/werc20/integration_test.go

```diff
@@ -0,0 +1,389 @@
+			// NOTE: Denom units MUST be increasing
+			DenomUnits: []*banktypes.DenomUnit{
+				{
+					Denom:    s.bondDenom,
+					Exponent: 0,
+					Aliases:  []string{"aevmos"},
+				},
+				{
+					Denom:    "aevmos",
+					Exponent: 18,
+				},
+			},
+			Name:    "Evmos",
+			Symbol:  "EVMOS",
+			Display: "aevmos",
+		}
+		s.network.App.BankKeeper.SetDenomMetaData(s.network.GetContext(), evmosMetadata)
+
+		tokenPair := erc20types.NewTokenPair(WEVMOSContractAddr, s.bondDenom, erc20types.OWNER_MODULE)
+
+		precompile, err := werc20.NewPrecompile(
+			tokenPair,
+			s.network.App.BankKeeper,
+			s.network.App.AuthzKeeper,
+			s.network.App.TransferKeeper,
+		)
+
+		Expect(err).ToNot(HaveOccurred(), "failed to create wevmos extension")
+		s.precompile = precompile
+
+		err = s.network.App.EvmKeeper.AddEVMExtensions(s.network.GetContext(), precompile)
+		Expect(err).ToNot(HaveOccurred(), "failed to add wevmos extension")
+
+		s.tokenDenom = tokenPair.GetDenom()
+
+		contractData = ContractData{
+			ownerPriv:      sender.Priv,
+			erc20Addr:      WEVMOSContractAddr,
+			erc20ABI:       testdata.WEVMOSContract.ABI,
+			precompileAddr: s.precompile.Address(),
+			precompileABI:  s.precompile.ABI,
+		}
+
+		failCheck = testutil.LogCheckArgs{ABIEvents: s.precompile.Events}
+		passCheck = failCheck.WithExpPass(true)
+
+		err = s.network.NextBlock()
+		Expect(err).ToNot(HaveOccurred(), "failed to advance block")
+
+		// Default sender and amount
+		sender = s.keyring.GetKey(0)
+		amount = big.NewInt(1e18)
+	})
+
+	Context("WEVMOS specific functions", func() {
+		When("calling deposit and withdraw correctly", func() {
+			It("calling deposit - should emit the Deposit event but not modify the balance", func() {
+				depositCheck := passCheck.WithExpPass(true).WithExpEvents(werc20.EventTypeDeposit)
+				txArgs, callArgs := s.getTxAndCallArgs(erc20Call, contractData, werc20.DepositMethod)
+				txArgs.Amount = amount
+
+				_, _, err := s.factory.CallContractAndCheckLogs(sender.Priv, txArgs, callArgs, depositCheck)
+				Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+				s.checkBalances(failCheck, sender, contractData)
+			})
+
+			It("calling withdraw - should emit the Withdrawal event but not modify the balance", func() {
+				withdrawCheck := passCheck.WithExpPass(true).WithExpEvents(werc20.EventTypeWithdrawal)
+				txArgs, callArgs := s.getTxAndCallArgs(erc20Call, contractData, werc20.WithdrawMethod, amount)
+
+				_, _, err := s.factory.CallContractAndCheckLogs(sender.Priv, txArgs, callArgs, withdrawCheck)
+				Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+				s.checkBalances(failCheck, sender, contractData)
+			})
+		})
+
+		// TODO: How do we actually check the method types here? We can see the correct ones being populated by printing the line in the cmn.Precompile
+		When("calling with incomplete data or amount", func() {
+			It("calls no call data, with amount - should call `receive` ", func() {
```

Comment:
please add tests for calling the same funcs (fallback) but without the amount defined (zero)